File: Difficulty: Medium/Egg Dropping Puzzle/egg-dropping-puzzle.py
class Solution:
    def eggDrop(self, n: int, k: int) -> int:
        # code here
        
        dp = [[-1] * (k + 1) for _ in range(n + 1)]
        def solve(n,k):
            if k==0: 
                return 0
            if n==1: 
                return k
                
            ans = float('inf')
            if dp[n][k]!=-1:
                return dp[n][k]
            low = 1
            high = k
            while low <= high:
                
                mid = (low + high)//2
                    
                eggbreaks = solve(n-1,mid-1)
                eggnotbreaks = solve(n,k-mid)
                
                worst = 1 + max(eggbreaks , eggnotbreaks)
                ans = min(ans,worst)
                
                if eggbreaks < eggnotbreaks:
                    low = mid+1
                else:
                    high = mid-1
            dp[n][k] =ans
            
            return dp[n][k]   
        return solve(n,k)
    

File: Difficulty: Medium/Egg Dropping Puzzle/egg-dropping-puzzle.cpp
class Solution {
  public:
  
   int solve(int n , int  k, vector<vector<int>>&dp)
   {
       if(k==0) return 0;
       if(n==1) return k;
       
       if(dp[n][k]!=-1) return dp[n][k];
       
       int low = 1;
       int high = k;
       int ans = INT_MAX;
       while(low<=high)
       {
           int mid = (low+high)/ 2;
           
           int breaks = solve(n-1,mid-1,dp);
           int not_breaks = solve(n,k-mid,dp);
           
           int worst = 1 + max(breaks,not_breaks);
           
           ans = min(ans,worst);
           
           if(breaks < not_breaks)
           {
               low = mid +1;
           }
           else
           {
               high = mid-1;
           }
       }
       
       return dp[n][k] = ans;
   }
    int eggDrop(int n, int k) {
        // code here
        
      vector<vector<int>> dp(n+1,vector(k+1,-1));
      
       return solve(n,k,dp);
    }
};